autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/importing/gpm/webflow/ImportGooglePasswordsWebFlowViewModel.kt

```diff
@@ -0,0 +1,131 @@
+import androidx.lifecycle.ViewModel
+import com.duckduckgo.anvil.annotations.ContributesViewModel
+import com.duckduckgo.autofill.impl.importing.CsvPasswordConverter
+import com.duckduckgo.autofill.impl.importing.CsvPasswordConverter.CsvPasswordImportResult
+import com.duckduckgo.autofill.impl.importing.CsvPasswordConverter.CsvPasswordImportResult.Success
+import com.duckduckgo.autofill.impl.importing.PasswordImporter
+import com.duckduckgo.autofill.impl.importing.gpm.webflow.ImportGooglePasswordResult.Companion.RESULT_KEY_DETAILS
+import com.duckduckgo.autofill.impl.importing.gpm.webflow.ImportGooglePasswordsWebFlowViewModel.ViewState.Initializing
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.di.scopes.ActivityScope
+import javax.inject.Inject
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import timber.log.Timber
+
+@ContributesViewModel(ActivityScope::class)
+class ImportGooglePasswordsWebFlowViewModel @Inject constructor() : ViewModel() {
+
+    @Inject
+    lateinit var dispatchers: DispatcherProvider
+
+    @Inject
+    lateinit var passwordImporter: PasswordImporter
+
+    @Inject
+    lateinit var csvPasswordConverter: CsvPasswordConverter
+
+    private val _viewState = MutableStateFlow<ViewState>(Initializing)
+    val viewState: StateFlow<ViewState> = _viewState
+
+    fun onPageStarted(url: String?) {
+        Timber.i("onPageStarted: $url")
+    }
+
+    fun onPageFinished(url: String?) {
+        Timber.i("onPageFinished: $url")
+    }
+
+    suspend fun onCsvAvailable(csv: String) {
+        when (val parseResult = csvPasswordConverter.readCsv(csv)) {
+            is CsvPasswordImportResult.Success -> onCsvParsed(parseResult)
+            is CsvPasswordImportResult.Error -> onCsvError()
+        }
+    }
+
+    private suspend fun onCsvParsed(parseResult: Success) {
+        val jobId =
+            passwordImporter.importPasswords(parseResult.loginCredentialsToImport, parseResult.numberPasswordsInSource)
+        val resultBundle = Bundle().also {
+            it.putParcelable(
+                RESULT_KEY_DETAILS,
+                ImportGooglePasswordResult.Success(
+                    importedCount = parseResult.loginCredentialsToImport.size,
+                    foundInImport = parseResult.loginCredentialsToImport.size,
+                    importJobId = jobId,
+                ),
+            )
+        }
+        _viewState.value = ViewState.UserFinishedImportFlow(resultBundle)
+    }
+
+    fun onCsvError() {
+        Timber.e("cdr Error decoding CSV")
+        val resultBundle = Bundle().also {
+            it.putParcelable(RESULT_KEY_DETAILS, ImportGooglePasswordResult.Error)
+        }
+        _viewState.value = ViewState.UserFinishedImportFlow(resultBundle)
+    }
+
+    fun onCloseButtonPressed(url: String?) {
+        if (url?.startsWith(ENCRYPTED_PASSPHRASE_ERROR_URL) == true) {
+            val resultBundle = Bundle().also {
```

Comment:
same, I would model the event agnostic from bundles here, and move the bundle to the view.